kernel/src/checkpoint/mod.rs

```diff
@@ -0,0 +1,485 @@
+        ADD_NAME, METADATA_NAME, PROTOCOL_NAME, REMOVE_NAME, SET_TRANSACTION_NAME, SIDECAR_NAME,
+    },
+    expressions::Scalar,
+    path::ParsedLogPath,
+    schema::{DataType, SchemaRef, StructField, StructType},
+    snapshot::Snapshot,
+    DeltaResult, Engine, EngineData, Error, EvaluationHandlerExtension,
+};
+use log_replay::{checkpoint_actions_iter, CheckpointData};
+use std::{
+    cell::RefCell,
+    rc::Rc,
+    sync::{Arc, LazyLock},
+    time::{Duration, SystemTime, UNIX_EPOCH},
+};
+use url::Url;
+
+mod log_replay;
+#[cfg(test)]
+mod tests;
+
+/// Schema for extracting relevant actions from log files during checkpoint creation
+static CHECKPOINT_READ_SCHEMA: LazyLock<SchemaRef> = LazyLock::new(|| {
+    StructType::new([
+        Option::<Add>::get_struct_field(ADD_NAME),
+        Option::<Remove>::get_struct_field(REMOVE_NAME),
+        Option::<Metadata>::get_struct_field(METADATA_NAME),
+        Option::<Protocol>::get_struct_field(PROTOCOL_NAME),
+        Option::<SetTransaction>::get_struct_field(SET_TRANSACTION_NAME),
+        Option::<Sidecar>::get_struct_field(SIDECAR_NAME),
+    ])
+    .into()
+});
+
+/// Returns the schema for reading Delta log actions during checkpoint creation
+fn get_checkpoint_read_schema() -> &'static SchemaRef {
+    &CHECKPOINT_READ_SCHEMA
+}
+
+/// Contains the path and data for a single-file checkpoint
+#[allow(unused)] // TODO(seb): Make pub for roll-out
+pub(crate) struct SingleFileCheckpointData {
+    /// Target URL where the checkpoint file will be written
+    pub path: Url,
+
+    /// Iterator over checkpoint actions to be written to the file
+    pub data: Box<dyn Iterator<Item = DeltaResult<CheckpointData>>>,
+}
+
+/// Manages the checkpoint writing process for Delta tables
+///
+/// The [`CheckpointWriter`] orchestrates creating checkpoint data and finalizing
+/// the checkpoint file. It tracks statistics about included actions and
+/// ensures checkpoint data is consumed only once.
+///
+/// # Usage Flow
+/// 1. Create via `Table::checkpoint()`
+/// 2. Call `get_checkpoint_info()` to obtain the [`SingleFileCheckpointData`]
+///    containing the path and action iterator for the checkpoint
+/// 3. Write the checkpoint data to storage (implementation-specific)
+/// 4. Call `finalize_checkpoint()` to create the _last_checkpoint file
+///
+/// # Internal Process
+/// 1. Reads relevant actions from the log segment using the checkpoint read schema
+/// 2. Applies selection and deduplication logic with the `CheckpointLogReplayProcessor`
+/// 3. Tracks counts of included actions for to be written to the _last_checkpoint file
+/// 5. Chains the [`CheckpointMetadata`] action to the actions iterator (for V2 checkpoints)
+#[allow(unused)] // TODO(seb): Make pub for roll-out
+pub(crate) struct CheckpointWriter {
+    /// The snapshot from which the checkpoint is created
+    pub(crate) snapshot: Snapshot,
```

Comment:
probably shouldn't own the snapshot? 
```suggestion
    pub(crate) snapshot: SnapshotRef,
```